testing/internal/e2e/tests/base/role_multi_scope_grants_test.go

```diff
@@ -0,0 +1,165 @@
+	e2e.MaybeSkipTest(t)
+	bc, err := boundary.LoadConfig()
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	boundary.AuthenticateAdminCli(t, ctx)
+
+	// Create Org and Project
+	newOrgId := boundary.CreateNewOrgCli(t, ctx)
+	t.Cleanup(func() {
+		ctx := context.Background()
+		boundary.AuthenticateAdminCli(t, ctx)
+		output := e2e.RunCommand(ctx, "boundary", e2e.WithArgs("scopes", "delete", "-id", newOrgId))
+		require.NoError(t, output.Err, string(output.Stderr))
+	})
+	newProjectId := boundary.CreateNewProjectCli(t, ctx, newOrgId)
+
+	// Create Account
+	acctName := "e2e-account"
+	newAccountId, acctPassword := boundary.CreateNewAccountCli(t, ctx, bc.AuthMethodId, acctName)
+	t.Cleanup(func() {
+		boundary.AuthenticateAdminCli(t, ctx)
+		output := e2e.RunCommand(ctx, "boundary",
+			e2e.WithArgs("accounts", "delete", "-id", newAccountId),
+		)
+		require.NoError(t, output.Err, string(output.Stderr))
+	})
+
+	// Create User and set Account to it
+	newUserId := boundary.CreateNewUserCli(t, ctx, "global")
+	t.Cleanup(func() {
+		boundary.AuthenticateAdminCli(t, ctx)
+		output := e2e.RunCommand(ctx, "boundary",
+			e2e.WithArgs("users", "delete", "-id", newUserId),
+		)
+		require.NoError(t, output.Err, string(output.Stderr))
+	})
+	boundary.SetAccountToUserCli(t, ctx, newUserId, newAccountId)
+
+	// Create Role and add admin grants to it
+	newRoleId, err := boundary.CreateRoleCli(t, ctx, "global")
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		boundary.AuthenticateAdminCli(t, ctx)
+		output := e2e.RunCommand(ctx, "boundary",
+			e2e.WithArgs("roles", "delete", "-id", newRoleId),
+		)
+		require.NoError(t, output.Err, string(output.Stderr))
+	})
+	boundary.AddGrantToRoleCli(t, ctx, newRoleId, "ids=*;type=*;actions=*")
+
+	// Authenticate test user and try to:
+	// - create role in global scope: expect error
+	// - create role in org scope: expect error
+	// - create role in proj scope: expect error
+	boundary.AuthenticateCli(t, ctx, bc.AuthMethodId, acctName, acctPassword)
+
+	_, err = boundary.CreateRoleCli(t, ctx, "global")
+	require.Error(t, err)
+
+	_, err = boundary.CreateRoleCli(t, ctx, newOrgId)
+	require.Error(t, err)
+
+	_, err = boundary.CreateRoleCli(t, ctx, newProjectId)
+	require.Error(t, err)
+
+	// Add User as a principle to the admin Role
+	boundary.AuthenticateAdminCli(t, ctx)
+	boundary.AddPrincipalToRoleCli(t, ctx, newRoleId, newUserId)
+
+	// Authenticate User and try to:
+	// - create Role in global scope: expect success
+	// - create Role in org scope: expect error
+	// - create Role in proj scope: expect error
+	boundary.AuthenticateCli(t, ctx, bc.AuthMethodId, acctName, acctPassword)
+	newGlobalRoleId2, err := boundary.CreateRoleCli(t, ctx, "global")
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		boundary.AuthenticateAdminCli(t, ctx)
+		output := e2e.RunCommand(ctx, "boundary",
+			e2e.WithArgs("roles", "delete", "-id", newGlobalRoleId2),
+		)
+		require.NoError(t, output.Err, string(output.Stderr))
+	})
```

Comment:
Curious if there's any other action we can use to check here besides creating a role (so that we don't have to clean up more stuff)

Reply:
I've chosen roles as it's a simple entity that can be created in each scope: global, org, proj.

We can get rid of two clean up calls if instead of attempting to create a role in global scope we'd attempt to update an existing one.

Then everything in org and proj scopes would be cleaned up by deleting the org.
And in global scope we'd need to delete the test-admin role only.

I feel that both approaches have its advantages; the alternative approach does save time but checks different operations (i.e. update, create) which does not feel 100% correct.

What do you think?

Reply:
> but checks different operations (i.e. update, create) which does not feel 100% correct.

Could you elaborate on this? Wouldn't it just be entirely dependent on the permissions we're granting? What if it was something like listing roles or something?

Reply:
> Could you elaborate on this?
I want the test to be doing a single operation, for example create role, in each scope under test.
I don't want the test to be doing one operation in global scope and another operation in org and proj scopes.

Listing roles will suffice as well.